rest/admin-v2/api/src/main/java/org/keycloak/models/mapper/MapStructClientModelMapper.java

```diff
@@ -1,46 +1,133 @@
+    @Mapping(target = "baseUrl", source = "appUrl")
+    @Mapping(target = "redirectUris", source = "appRedirectUrls")
+    @Mapping(target = "authenticationFlowBindingOverrides", source = "loginFlows", ignore = true) // TODO
+    @Mapping(target = "publicClient", source = "auth.enabled", qualifiedByName = "isPublicClientPrimitive")
+    @Mapping(target = "clientAuthenticatorType", source = "auth.method")
+    @Mapping(target = "secret", source = "auth.secret")
+    @Mapping(target = "serviceAccountsEnabled", source = "serviceAccount.enabled")
+    @interface RepToModel {
+    }
+
     @Mapping(target = "displayName", source = "name")
     @Mapping(target = "appUrl", source = "baseUrl")
     @Mapping(target = "appRedirectUrls", source = "redirectUris")
     @Mapping(target = "loginFlows", source = "authenticationFlowBindingOverrides", ignore = true)
-    @Mapping(target = "auth", ignore = true) // TODO
+    @Mapping(target = "auth.enabled", source = "publicClient", qualifiedByName = "isPublicClient")
+    @Mapping(target = "auth.method", source = "clientAuthenticatorType")
+    @Mapping(target = "auth.secret", source = "secret")
+    @Mapping(target = "auth.certificate", ignore = true) // no cert in the representation
     @Mapping(target = "roles", source = "rolesStream", qualifiedByName = "getRoleStrings")
     @Mapping(target = "serviceAccount.enabled", source = "serviceAccountsEnabled")
-    @Mapping(target = "serviceAccount.roles", source = "rolesStream", qualifiedByName = "getServiceAccountRoles")
-    @Override
-    ClientRepresentation fromModel(ClientModel model);
+    @Mapping(target = "serviceAccount.roles", source = ".", qualifiedByName = "getServiceAccountRoles")
+    @interface ModelToRep {
+    }
 
-    // we don't want to ignore nulls so that we completely overwrite the state
-    @Override
-    void toModel(@MappingTarget ClientModel model, ClientRepresentation rep, @Context RealmModel realm);
+    /*-------------------------------------*
+     *          HELPER METHODS             *
+     *-------------------------------------*/
+    @ObjectFactory
+    default ClientModel createClientModel(@Context RealmModel realm, ClientRepresentation rep) {
+        // dummy add/remove to obtain a detached model
+        var model = realm.addClient(rep.getClientId());
+        realm.removeClient(model.getId());
+        return model;
+    }
+
+    @AfterMapping
+    default void addRoles(@MappingTarget ClientModel model, ClientRepresentation rep, @Context RealmModel realm, @Context KeycloakSession session) {
+        Optional.ofNullable(rep.getRoles())
+                .orElse(Collections.emptySet())
+                .stream()
+                .filter(role -> model.getRole(role) == null)
+                .forEach(model::addRole);
+
+        // Service Account roles
+        var serviceAccount = rep.getServiceAccount();
+        if (serviceAccount != null && serviceAccount.getEnabled() && !serviceAccount.getRoles().isEmpty()) {
```

Comment:
I assume this won't remove roles and disable the service account on update, right? For now, we should probably just leverage the v1 logic here...